src/Workspaces/Core/MSBuild/MSBuild/Build/ProjectBuildManager.cs

```diff
@@ -0,0 +1,296 @@
+
+        private readonly MSB.Evaluation.ProjectCollection _projectCollection;
+        private readonly MSBuildDiagnosticLogger _logger;
+        private bool _started;
+
+        public ProjectBuildManager()
+        {
+            var properties = new Dictionary<string, string>()
+            {
+                { "DesignTimeBuild", "true" }, // this will tell msbuild to not build the dependent projects
+                { "BuildingInsideVisualStudio", "true" }, // this will force CoreCompile task to execute even if all inputs and outputs are up to date
+                { "BuildProjectReferences", "false" },
+                { "BuildingProject", "false" },
+                { "ProvideCommandLineArgs", "true" }, // retrieve the command-line arguments to the compiler
+                { "SkipCompilerExecution", "true" }, // don't actually run the compiler
+                { "ContinueOnError", "ErrorAndContinue" }
+            };
+
+            _projectCollection = new MSB.Evaluation.ProjectCollection(properties);
+
+            _logger = new MSBuildDiagnosticLogger()
+            {
+                Verbosity = MSB.Framework.LoggerVerbosity.Normal
+            };
+        }
+
+        private MSB.Evaluation.Project FindProject(string path, IDictionary<string, string> globalProperties)
+        {
+            var loadedProjects = _projectCollection.GetLoadedProjects(path);
+            if (loadedProjects == null || loadedProjects.Count == 0)
+            {
+                return null;
+            }
+
+            globalProperties = globalProperties ?? ImmutableDictionary<string, string>.Empty;
+            var totalGlobalProperties = _projectCollection.GlobalProperties.Count + globalProperties.Count;
+
+            foreach (var loadedProject in loadedProjects)
+            {
+                if (loadedProject.GlobalProperties.Count != totalGlobalProperties)
+                {
+                    continue;
+                }
+
+                // All projects in the collection should have the default global properties, so
+                // there's no need to check collection.GlobalProperties.
+
+                var found = true;
+                foreach (var globalProp in globalProperties)
+                {
+                    if (!loadedProject.GlobalProperties.Contains(globalProp))
+                    {
+                        found = false;
+                        break;
+                    }
+                }
+
+                if (found)
+                {
+                    return loadedProject;
+                }
+            }
+
+            return null;
+        }
+
+        public async Task<(MSB.Evaluation.Project project, DiagnosticLog log)> LoadProjectAsync(
+            string path, IDictionary<string, string> globalProperties, CancellationToken cancellationToken)
+        {
+            var log = new DiagnosticLog();
+
+            try
+            {
+                var project = FindProject(path, globalProperties);
```

Comment:
since FindProject has loop, giving in cancellation token seems good thing to do.

Reply:
Sure. Will do.